libtenzir/builtins/operators/from_to_2.cpp

```diff
@@ -582,9 +645,702 @@ class to_plugin2 final : public virtual operator_factory_plugin {
+    auto& format = compression_and_format.format.get();
+    auto compression = compression_and_format.compression;
+    auto provider = session_provider::make(path_dh);
+    auto ctx = provider.as_session();
+    auto pipe = pipeline{};
+    if (compression) {
+      TRY(auto decompress, make_operator(*compression, args_.url.source, ctx));
+      pipe.append(std::move(decompress));
+    }
+    TRY(auto read, make_operator(format, args_.url.source, ctx));
+    pipe.append(std::move(read));
+    return pipe;
+  }
+
+  void start_job(const arrow::fs::FileInfo& file) {
+    TENZIR_ASSERT(active_jobs_ < max_jobs);
+    active_jobs_ += 1;
+    auto pipe = make_pipeline(file.path());
+    if (pipe.is_error()) {
+      active_jobs_ -= 1;
+      return;
+    }
+    // We already checked the output type after parsing.
+    auto output_type = pipe->infer_type<chunk_ptr>();
+    TENZIR_ASSERT(output_type);
+    TENZIR_ASSERT(output_type->is<table_slice>());
+    add_actor_callback(
+      self_, fs_->OpenInputStreamAsync(file),
+      [this, pipe = std::move(*pipe), path = file.path()](
+        arrow::Result<std::shared_ptr<arrow::io::InputStream>> stream) mutable {
+        if (not stream.ok()) {
+          diagnostic::error("failed to open `{}`", path)
+            .primary(args_.url)
+            .note(stream.status().ToStringWithoutContextLines())
+            .emit(*dh_);
+          active_jobs_ -= 1;
+          return;
+        }
+        auto source = self_->spawn(caf::actor_from_state<arrow_chunk_source>,
+                                   std::move(*stream));
+        auto weak = caf::weak_actor_ptr{source->ctrl()};
+        pipe.prepend(std::make_unique<from_file_source>(std::move(source)));
+        pipe.append(std::make_unique<from_file_sink>(
+          self_, order_,
+          args_.path_field ? std::optional{std::pair{*args_.path_field, path}}
+                           : std::nullopt));
+        pipe = pipe.optimize_if_closed();
+        auto executor
+          = self_->spawn(pipeline_executor, std::move(pipe), definition_, self_,
+                         self_, node_, false, is_hidden_);
+        self_->attach_functor([this, weak]() {
+          if (auto strong = weak.lock()) {
+            // FIXME: This should not be necessary to ensure that the actor is
+            // destroyed when the executor is done. This problem could also
+            // apply to other operators.
+            self_->send_exit(strong, caf::exit_reason::user_shutdown);
+          }
+        });
+        self_->monitor(executor, [this, executor, path,
+                                  weak = std::move(weak)](caf::error error) {
+          if (auto strong = weak.lock()) {
+            // FIXME: This should not be necessary to ensure that the actor is
+            // destroyed when the executor is done. This problem could also
+            // apply to other operators.
+            self_->send_exit(strong, caf::exit_reason::user_shutdown);
+          }
```

Comment:
Let's talk through this once tomorrow.

Reply:
I'll just leave this in to be safe, okay? Or do you want me to spend the time checking whether our latest changes also make this work in all scenarios?